locust/exporter.py

```diff
@@ -0,0 +1,141 @@
+
+        self._client = InfluxDBClient(
+            url=INFLUXDB_URL,
+            token=INFLUXDB_URL,
+            org=INFLUXDB_ORG,
+        )
+        self._BucketsApi = BucketsApi
+        self.stat_entries = {}
+        self._points = []
+        self._environment = environment
+        self._environment.events.request.add_listener(self.on_request)
+        self._environment.events.test_start.add_listener(self.on_test_start)
+        self._environment.events.test_stop.add_listener(self.on_test_stop)
+        self._write_api = self._client.write_api(write_options=SYNCHRONOUS)
+        self._query_api = self._client.query_api()
+
+        self._greenlets = []
+
+        atexit.register(self.on_exit)
+
+    def _create_bucket(self):
+        buckets_api = self._BucketsApi(self._client)
+        current_time = datetime.now().strftime("%Y-%m-%d_%H-%M")
+        bucket_name = f"locust_test_{current_time}"
+
+        try:
+            bucket = buckets_api.create_bucket(bucket_name=bucket_name)
+            return bucket.name
+        except Exception as e:
+            print(e)
+            sys.stderr.write(
+                "Could not create InfluxDB bucket. Please ensure the INFLUXDB_URL has access to create new buckets.\n"
+            )
+            sys.exit(1)
+
+    def _queue_point(self, measurement, tags={}, fields={}):
+        point = {"measurement": measurement, "tags": tags, "fields": fields, "time": datetime.utcnow().isoformat()}
+        self._points.append(point)
+
+    def _write_points(self):
+        while True:
+            # Copy and clear queue so only new points are sent each cycle
+            points = self._points.copy()
+            self._points.clear()
+
+            if points:
+                self._write_api.write(record=points, bucket=self._bucket)
+
+            gevent.sleep(0.5)
+
+    def _log_user_count(self):
+        while True:
+            user_count = self._environment.runner.user_count
+
+            self._queue_point(
+                "user_count",
+                fields={"user_count": user_count},
+            )
+
+            gevent.sleep(2)
+
+    def on_exit(self):
+        for greenlet in self._greenlets:
+            greenlet.kill()
+
+    def register(self, stat_entries):
+        self.stat_entries = stat_entries
+
+    def on_test_start(self, environment):
+        self._bucket = self._create_bucket()
+
+        if environment.runner and not isinstance(environment.runner, WorkerRunner):
+            self._greenlets.append(gevent.spawn(self._log_user_count))
+
+        self._greenlets.append(gevent.spawn(self._write_points))
+
+    def on_test_stop(self, environment):
+        self.on_exit()
+
+    def on_request(
+        self,
+        request_type,
+        name,
+        response_time,
+        response_length,
+        exception,
+        context,
+        start_time=None,
+        url=None,
+        **kwargs,
+    ):
+        measurement = "request_failure" if exception else "request_success"
```

Comment:
I'm less versed then you are in InfluxDB, but shouldnt it be just _one_ measurement (request)? Wouldnt that make queries etc much easier?

Reply:
It helps us to separate the failed requests versus the successful ones, for example, if we wanted to plot request_failures / s we could use a query like so:
```
from(bucket: v.bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_field"] == "response_time" and r["_measurement"] == "request_failure")
```

Reply:
Yes, but what about the (slightly more common) queries where we want *all* requests? At least that is how I would do it in SQL...

Reply:
```
from(bucket: v.bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_field"] == "response_time")
```
Would give back all requests
But yeah on the other hand, then we could also count errors like
```
from(bucket: v.bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_field"] == "exception")
```
So I think it's just a matter of preference

Reply:
Just reading the docs (https://docs.influxdata.com/influxdb/cloud/write-data/best-practices/schema-design/) I _think_ we should store a tag `success` (with true/false value). But lets chat about it on monday.

Reply:
Yeah so based on the docs it definitely makes sense to just have the one measurement. Whether or not we want to have the tag `success`, we should decide if it's useful. Like I mentioned, since we could query for failures with `r["_field"] == "exception"`, we don't necessarily need a `success` field

Reply:
Don't we need it to be a tag for faster querying? Fields arent indexed, right?

Reply:
Yup sorry! For some reason I was thinking exception was a tag but obviously it's not